### PR TITLE
[WIP] Migrating F1 Score

### PR DESCRIPTION
**Pull request by SSaishruthi**
This PR is for migrating F1 score from addons to the core.

@pavithrasv will be reviewing the PR
@seanpmorgan @karmel  for your reference

### CODE REVIEW

**Comment by SSaishruthi** [comment]
Did not add test scripts yet. Will do after the review comments

**Comment by SSaishruthi** [comment]
Hi @pavithrasv 
Did you get a chance to review the PR?

File: tensorflow/python/keras/metrics.py [hunk 1]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+
+    `beta < 1` gives more weight to the precision.
+    `beta > 1` gives more weight to the recall.
+    `beta == 1` gives equal weight to precision and recall.
+
+    Args:
+       num_classes : Number of unique classes in the dataset.
+       average : Type of averaging to be performed on data.
+                   Acceptable values are None, micro, macro and
+                   weighted.
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: Are these strings? Could we use quotes around strings for all the usages to make it clear?

File: tensorflow/python/keras/metrics.py [hunk 2]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+    precision and recall.
+
+    `beta < 1` gives more weight to the precision.
+    `beta > 1` gives more weight to the recall.
+    `beta == 1` gives equal weight to precision and recall.
+
+    Args:
+       num_classes : Number of unique classes in the dataset.
+       average : Type of averaging to be performed on data.
+                   Acceptable values are None, micro, macro and
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: Could we indent these 2 spaces from where average starts? Also for other usages below.

File: tensorflow/python/keras/metrics.py [hunk 3]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+    4. If `weighted` is specified, metrics are computed for
+       each class and returns the mean weighted by the
+       number of true instances in each class taking data
+       imbalance into account.
+
+    Usage:
+    ```python
+    actuals = tf.constant([[1, 1, 0],[1, 0, 0]],
+              dtype=tf.int32)
+    predis = tf.constant([[1, 0, 0],[1, 0, 1]],
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: In general there are a lot of formatting issues, please wrap to next line if > 80 chars and indent continuation line using 4 spaces.

File: tensorflow/python/keras/metrics.py [hunk 4]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+             dtype=tf.int32)
+    # F-Beta Micro
+    fb_score = tfa.metrics.FBetaScore(num_classes=3,
+                beta=0.4, average='micro')
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: tf.metrics

File: tensorflow/python/keras/metrics.py [hunk 5]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+        super(FBetaScore, self).__init__(name=name)
+        self.num_classes = num_classes
+        # type check
+        if not isinstance(beta, float) and beta.dtype != tf.float32:
+            raise TypeError("The value of beta should be float")
+        # value check
+        if beta <= 0.0:
+            raise ValueError("beta value should be greater than zero")
+        else:
+            self.beta = beta
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: Please use ' for quotation to be consistent with the rest of the file.

File: tensorflow/python/keras/metrics.py [hunk 6]
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
...
+                dtype=self.dtype)
+            self.weights_intermediate = self.add_weight(
+                'weights',
+                shape=[self.num_classes],
+                initializer='zeros',
+                dtype=self.dtype)
+
+    # TO DO SSaishruthi: Add sample weight option
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        y_true = tf.cast(y_true, tf.int32)
```
Comment (pavithrasv) [inline tensorflow/python/keras/metrics.py]: Please raise an exception for unsupported use cases.

**Review by pavithrasv — COMMENTED** [review]
Thank you for the PR. Did a quick review, in general i am seeing a lot of formatting issues, please refer to the code in the rest of the file for formatting details and also please add comprehensive unit tests for the metric.

**Comment by SSaishruthi** [comment]
Thanks @pavithrasv 
Do you have any sample test scripts that I can refer to? I have written one for addons and will combine both.

**Comment by SSaishruthi** [comment]
Hi @pavithrasv 

I have updated the following:

1.  Corrected lines > 80 characters.
2. Changed all `tf.math` to `math_ops` and `tf.reduce_sum` to `math_ops.reduce_sum`
3. Added usage for Keras API
4. Changed `tfa.metrics` to `tf.keras.metrics`.
5. Changed indentation for docstrings

Clarification:

1. We are not exposing `sample_weight` parameter now. Do we need to add exception for the same?
2. I have test cases for F1 in addons. Can we modify the same or I need to refer some existing sample scripts?

https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/f1_test.py


Please let me know if the current changes look good

**Comment by SSaishruthi** [comment]
I will be adding a `threshold` parameter today. Will ping again once I update

**Comment by PhilipMay** [comment]
There are still issues in the implementation. See here: https://github.com/tensorflow/addons/issues/490

**Comment by SSaishruthi** [comment]
@pavithrasv The issue has been taken care. we are good for now

**Comment by SSaishruthi** [comment]
Updating scripts

**Comment by SSaishruthi** [comment]
Waiting for the f1 score update PR to get merged. Will update the scripts right after that

**Comment by SSaishruthi** [comment]
We have an updated version of the code. It is almost ready. I will update the code here by tomorrow and we can take it forward from here.

**Comment by SSaishruthi** [comment]
Refractor PR got merged. Now, we are good to go with migration.
Updated new scripts. Need to do some minor formatting work.

**Comment by pavithrasv** [comment]
You can add tests in the file metrics_test.py

**Comment by SSaishruthi** [comment]
Thanks, @pavithrasv will update.

**Comment by SSaishruthi** [comment]
I have added a test. Will do a final cleanup in a day and we can have a final review on this.
Thanks

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by SSaishruthi** [comment]
This is WIP. Will be completed this week

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by PhilipMay** [comment]
F1 metric in addons still seems to have a very ugly bug: https://github.com/tensorflow/addons/issues/825

**Comment by gbaned** [comment]
@SSaishruthi Any update on this PR, please. Thanks!

**Comment by SSaishruthi** [comment]
@gbaned Hi

Sorry I was on vacation. We have some clarification to make. Will keep till we clear that

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@SSaishruthi Any update on this PR and can you please resolve conflicts? Thanks!

**Comment by PhilipMay** [comment]
@gbaned No offence but I think @SSaishruthi does not have the time at the moment. Someone else should pick up the migration work. The TF Addons implementation seems to be ok. Want me to create a new PR for this migration?

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
